ext/mysqli/mysqli_api.c

```diff
@@ -819,15 +819,43 @@ PHP_FUNCTION(mysqli_stmt_execute)
 {
 	MY_STMT		*stmt;
 	zval		*mysql_stmt;
+	zval		*input_params = NULL;
 #ifndef MYSQLI_USE_MYSQLND
 	unsigned int	i;
 #endif
 
-	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "O", &mysql_stmt, mysqli_stmt_class_entry) == FAILURE) {
+	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "O|a", &mysql_stmt, mysqli_stmt_class_entry, &input_params) == FAILURE) {
 		RETURN_THROWS();
 	}
 	MYSQLI_FETCH_RESOURCE_STMT(stmt, mysql_stmt, MYSQLI_STATUS_VALID);
 
+	// bind-in-execute
+#if defined(MYSQLI_USE_MYSQLND)
+	if(input_params) {
+		zval *tmp;
+		unsigned int index;
+		MYSQLND_PARAM_BIND	*params;
+
+		params = mysqlnd_stmt_alloc_param_bind(stmt->stmt);
+		if (!params) {
+			// can we safely return here?
+			RETVAL_FALSE;
+		}
+
+		index = 0;
+		ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(input_params), tmp) {
```

Comment:
I'm not sure. I based it off PDO_stmt::execute(). My test cases passed, but if you think we should add more checks here, please suggest.

Reply:
Adding a test case that passes too many parameters should illustrate the issue. I expect that it will result in memory errors.

You need to check that `zend_hash_num_elements(Z_ARRVAL_P(input_params)) == mysql_stmt_param_count(stmt->stmt)` and throw a ValueError otherwise.

Reply:
The problem here is that mysqlnd doesn't complain about too many params. If you try to bind more than necessary then the rest is simply ignored. 

I have added an ArgumentCountError if more than needed values are present but my problem with this is that it is not a client error but it should be. It should be mysqli_sql_exception, a warning or it should be possible to silence it. However, if it is a hard error like this then the behaviour will be reported as inconsistent. Also, ArgumentCountError is not the right choice, given that the number of arguments from user POV matches. 

So, I went with an option where we ignore the client errors, and instead just throw a generic ValueError for all cases where the number of values does not match the number of params. This is at least going to be consistent with the function itself and the `bind_param` method.

I believe this is overall an improvement, so thanks for the suggestion.